Deploys sometimes fail without properly cleaning up

There have been a couple of cases where a deploy has failed but not cleaned up properly causing subsequent deploys of the same stack/stage to wait forever and the inability to deploy riff-raff itself due to it thinking there are on-going deploys.

A couple of examples:
94d5f1fe-7b44-4d8b-b11c-5be4a8716045
95f35c99-c7ce-47c5-add5-f3a6eccc7670

For the record here are some log snippets.

For `95f35c99-c7ce-47c5-add5-f3a6eccc7670`:
```text
2017-05-11 14:31:53,932 [application-akka.actor.default-dispatcher-3620] INFO  d.Deployments - Started deploying DeployParameters(Deployer(Philip McMahon),Build(Editorial Tools::Workflow Frontend,289),Stage(CODE),RecipeName(default),List(),List(),All)
2017-05-11 14:31:53,937 [application-akka.actor.default-dispatcher-3620] INFO  d.Deployments - Creating deploy record for DeployParameters(Deployer(Philip McMahon),Build(Editorial Tools::Workflow Frontend,289),Stage(CODE),RecipeName(default),List(),List(),All)
2017-05-11 14:31:53,969 [application-akka.actor.default-dispatcher-3620] DEBUG d.DeploymentEngine - Sending start deploy message to co-ordinator
2017-05-11 14:31:53,969 [deploy-akka.actor.default-dispatcher-4] DEBUG d.a.DeployCoordinator - Scheduling deploy
2017-05-11 14:31:54,037 [deploy-akka.deploy-dispatcher-1470] DEBUG d.a.DeployGroupRunner - Running next set of tasks (workflow-frontend [uploadArtifacts] => eu-west-1/workflow/0) on actor 95f35c99-c7ce-47c5-add5-f3a6eccc7670-1
2017-05-11 14:31:54,037 [deploy-akka.deploy-dispatcher-1470] DEBUG d.a.DeployGroupRunner - Running next set of tasks (workflow-frontend-ami-update [update] => eu-west-1/workflow/1) on actor 95f35c99-c7ce-47c5-add5-f3a6eccc7670-2
2017-05-11 14:31:54,037 [deploy-akka.deploy-dispatcher-1470] DEBUG d.actors.TasksRunner - New tasks runner created with path akka://deploy/user/deployCoordinator/deployGroupRunner-95f35c99-c7ce-47c5-add5-f3a6eccc7670/deploymentRunner-95f35c99-c7ce-47c5-add5-f3a6eccc7670-1
2017-05-11 14:31:54,037 [deploy-akka.deploy-dispatcher-1470] DEBUG d.actors.TasksRunner - Running task workflow-frontend [uploadArtifacts] => eu-west-1/workflow/0
2017-05-11 14:31:54,037 [deploy-akka.deploy-dispatcher-1471] DEBUG d.actors.TasksRunner - New tasks runner created with path akka://deploy/user/deployCoordinator/deployGroupRunner-95f35c99-c7ce-47c5-add5-f3a6eccc7670/deploymentRunner-95f35c99-c7ce-47c5-add5-f3a6eccc7670-2
2017-05-11 14:31:54,037 [deploy-akka.deploy-dispatcher-1471] DEBUG d.actors.TasksRunner - Running task workflow-frontend-ami-update [update] => eu-west-1/workflow/0
2017-05-11 14:31:54,040 [deploy-akka.deploy-dispatcher-1472] DEBUG m.tasks.S3Upload - Transferring s3://riffraff-artifact/Editorial Tools::Workflow Frontend/289/workflow-frontend/workflow-frontend.tgz to s3://workflow-frontend/workflow/CODE/workflow-frontend/workflow-frontend.tgz with CacheControl:None ContentType:None PublicRead:false
2017-05-11 14:31:54,071 [deploy-akka.deploy-dispatcher-1472] WARN  c.a.s.s.i.S3AbortableInputStream - Not all bytes were read from the S3ObjectInputStream, aborting HTTP connection. This is likely an error and may result in sub-optimal behavior. Request only the bytes you need via a ranged GET or drain the input stream after use.
2017-05-11 14:31:54,071 [deploy-akka.deploy-dispatcher-1472] WARN  c.a.s.s.i.S3AbortableInputStream - Not all bytes were read from the S3ObjectInputStream, aborting HTTP connection. This is likely an error and may result in sub-optimal behavior. Request only the bytes you need via a ranged GET or drain the input stream after use.
2017-05-11 14:31:54,071 [deploy-akka.deploy-dispatcher-1472] WARN  c.a.s.s.i.S3AbortableInputStream - Not all bytes were read from the S3ObjectInputStream, aborting HTTP connection. This is likely an error and may result in sub-optimal behavior. Request only the bytes you need via a ranged GET or drain the input stream after use.
2017-05-11 14:31:54,071 [deploy-akka.deploy-dispatcher-1472] WARN  c.a.s.s.i.S3AbortableInputStream - Not all bytes were read from the S3ObjectInputStream, aborting HTTP connection. This is likely an error and may result in sub-optimal behavior. Request only the bytes you need via a ranged GET or drain the input stream after use.
2017-05-11 14:31:54,071 [deploy-akka.deploy-dispatcher-1472] WARN  c.a.s.s.i.S3AbortableInputStream - Not all bytes were read from the S3ObjectInputStream, aborting HTTP connection. This is likely an error and may result in sub-optimal behavior. Request only the bytes you need via a ranged GET or drain the input stream after use.
2017-05-11 14:31:54,071 [deploy-akka.deploy-dispatcher-1472] WARN  c.a.s.s.i.S3AbortableInputStream - Not all bytes were read from the S3ObjectInputStream, aborting HTTP connection. This is likely an error and may result in sub-optimal behavior. Request only the bytes you need via a ranged GET or drain the input stream after use.
2017-05-11 14:31:54,071 [deploy-akka.deploy-dispatcher-1472] WARN  c.a.s.s.i.S3AbortableInputStream - Not all bytes were read from the S3ObjectInputStream, aborting HTTP connection. This is likely an error and may result in sub-optimal behavior. Request only the bytes you need via a ranged GET or drain the input stream after use.
2017-05-11 14:31:54,072 [deploy-akka.deploy-dispatcher-1472] WARN  c.a.s.s.i.S3AbortableInputStream - Not all bytes were read from the S3ObjectInputStream, aborting HTTP connection. This is likely an error and may result in sub-optimal behavior. Request only the bytes you need via a ranged GET or drain the input stream after use.
2017-05-11 14:31:54,072 [deploy-akka.deploy-dispatcher-1472] WARN  magenta.tasks.S3$ - Encountered fatal exception during AWS API call
com.amazonaws.services.s3.model.AmazonS3Exception: The specified bucket does not exist (Service: Amazon S3; Status Code: 404; Error Code: NoSuchBucket; Request ID: C7B550EE3D463998)
	at com.amazonaws.http.AmazonHttpClient$RequestExecutor.handleErrorResponse(AmazonHttpClient.java:1588)
	at com.amazonaws.http.AmazonHttpClient$RequestExecutor.executeOneRequest(AmazonHttpClient.java:1258)
	at com.amazonaws.http.AmazonHttpClient$RequestExecutor.executeHelper(AmazonHttpClient.java:1030)
	at com.amazonaws.http.AmazonHttpClient$RequestExecutor.doExecute(AmazonHttpClient.java:742)
	at com.amazonaws.http.AmazonHttpClient$RequestExecutor.executeWithTimer(AmazonHttpClient.java:716)
	at com.amazonaws.http.AmazonHttpClient$RequestExecutor.execute(AmazonHttpClient.java:699)
	at com.amazonaws.http.AmazonHttpClient$RequestExecutor.access$500(AmazonHttpClient.java:667)
	at com.amazonaws.http.AmazonHttpClient$RequestExecutionBuilderImpl.execute(AmazonHttpClient.java:649)
	at com.amazonaws.http.AmazonHttpClient.execute(AmazonHttpClient.java:513)
	at com.amazonaws.services.s3.AmazonS3Client.invoke(AmazonS3Client.java:4187)
	at com.amazonaws.services.s3.AmazonS3Client.invoke(AmazonS3Client.java:4134)
	at com.amazonaws.services.s3.AmazonS3Client.putObject(AmazonS3Client.java:1720)
	at magenta.tasks.S3Upload$$anonfun$execute$1$$anonfun$apply$1.apply(tasks.scala:69)
	at magenta.tasks.S3Upload$$anonfun$execute$1$$anonfun$apply$1.apply(tasks.scala:65)
	at scala.util.Try$.apply(Try.scala:192)
	at magenta.package$.retryOnException(package.scala:40)
	at magenta.tasks.S3Upload$$anonfun$execute$1.apply(tasks.scala:65)
	at magenta.tasks.S3Upload$$anonfun$execute$1.apply(tasks.scala:58)
	at scala.collection.Iterator$class.foreach(Iterator.scala:893)
	at scala.collection.AbstractIterator.foreach(Iterator.scala:1336)
2017-05-11 14:31:54,072 [deploy-akka.deploy-dispatcher-1470] DEBUG d.actors.TasksRunner - Sending failed message
2017-05-11 14:31:54,072 [deploy-akka.deploy-dispatcher-1470] DEBUG d.a.DeployGroupRunner - Deployment failed
2017-05-11 14:31:54,072 [deploy-akka.deploy-dispatcher-1470] DEBUG d.a.DeployGroupRunner - Failed during deployment but others still running - deferring clean up
2017-05-11 14:32:00,699 [deploy-akka.deploy-dispatcher-1471] DEBUG d.actors.TasksRunner - Running task workflow-frontend-ami-update [update] => eu-west-1/workflow/1
2017-05-11 14:32:04,395 [RxComputationScheduler-1] INFO  ci.CIBuild - Found S3Build(962,editorial-tools:tag-manager,editorial-tools:tag-manager,microsite-notes,962,2017-05-11T14:31:20.137Z,f42d21881a8412a9b1763e8229a9c28524e72243,git@github.com:guardian/tagmanager.git)
2017-05-11 14:32:04,396 [RxComputationScheduler-1] INFO  ci.CIBuild - Found S3Build(962,editorial-tools:tag-manager,editorial-tools:tag-manager,microsite-notes,962,2017-05-11T14:31:20.137Z,f42d21881a8412a9b1763e8229a9c28524e72243,git@github.com:guardian/tagmanager.git)
2017-05-11 14:32:04,396 [RxComputationScheduler-1] INFO  ci.CIBuild - Found S3Build(962,editorial-tools:tag-manager,editorial-tools:tag-manager,microsite-notes,962,2017-05-11T14:31:20.137Z,f42d21881a8412a9b1763e8229a9c28524e72243,git@github.com:guardian/tagmanager.git)
2017-05-11 14:32:14,249 [RxComputationScheduler-1] INFO  ci.CIBuild - Found S3Build(1746,ophan::Data Lake::ophan-event-stream-etl,ophan::Data Lake::ophan-event-stream-etl,pull/794,1746,2017-05-11T14:28:41.960Z,3f885d4ff729f2d43d0bc9a76077f8c2f349d332,git@github.com:guardian/ophan-data-lake.git)
2017-05-11 14:32:14,249 [RxComputationScheduler-1] INFO  ci.CIBuild - Found S3Build(1746,ophan::Data Lake::ophan-event-stream-etl,ophan::Data Lake::ophan-event-stream-etl,pull/794,1746,2017-05-11T14:28:41.960Z,3f885d4ff729f2d43d0bc9a76077f8c2f349d332,git@github.com:guardian/ophan-data-lake.git)
2017-05-11 14:32:14,249 [RxComputationScheduler-1] INFO  ci.CIBuild - Found S3Build(1746,ophan::Data Lake::ophan-event-stream-etl,ophan::Data Lake::ophan-event-stream-etl,pull/794,1746,2017-05-11T14:28:41.960Z,3f885d4ff729f2d43d0bc9a76077f8c2f349d332,git@github.com:guardian/ophan-data-lake.git)
2017-05-11 14:32:35,666 [application-akka.actor.default-dispatcher-3613] WARN  o.m.d.connection - Got socket exception on connection [connectionId{localValue:303}] to aws-eu-west-1-portal.0.dblayer.com:11301. All connections to aws-eu-west-1-portal.0.dblayer.com:11301 will be closed.
2017-05-11 14:32:35,666 [application-akka.actor.default-dispatcher-3613] INFO  o.m.d.connection - Closed connection [connectionId{localValue:303}] to aws-eu-west-1-portal.0.dblayer.com:11301 because there was a socket exception raised by this connection.
2017-05-11 14:32:35,666 [application-akka.actor.default-dispatcher-3613] ERROR p.MongoDatastore - Squashing uncaught exceptionwhilst Counting documents matching filter
com.mongodb.MongoSocketReadException: Exception receiving message
	at com.mongodb.connection.InternalStreamConnection.translateReadException(InternalStreamConnection.java:480)
	at com.mongodb.connection.InternalStreamConnection.receiveMessage(InternalStreamConnection.java:225)
	at com.mongodb.connection.UsageTrackingInternalConnection.receiveMessage(UsageTrackingInternalConnection.java:102)
	at com.mongodb.connection.DefaultConnectionPool$PooledConnection.receiveMessage(DefaultConnectionPool.java:433)
	at com.mongodb.connection.CommandProtocol.execute(CommandProtocol.java:109)
	at com.mongodb.connection.DefaultServer$DefaultServerProtocolExecutor.execute(DefaultServer.java:159)
	at com.mongodb.connection.DefaultServerConnection.executeProtocol(DefaultServerConnection.java:286)
	at com.mongodb.connection.DefaultServerConnection.command(DefaultServerConnection.java:173)
	at com.mongodb.operation.CommandOperationHelper.executeWrappedCommandProtocol(CommandOperationHelper.java:215)
	at com.mongodb.operation.CommandOperationHelper.executeWrappedCommandProtocol(CommandOperationHelper.java:206)
	at com.mongodb.operation.CommandOperationHelper.executeWrappedCommandProtocol(CommandOperationHelper.java:112)
	at com.mongodb.operation.CountOperation$1.call(CountOperation.java:210)
	at com.mongodb.operation.CountOperation$1.call(CountOperation.java:206)
	at com.mongodb.operation.OperationHelper.withConnectionSource(OperationHelper.java:230)
	at com.mongodb.operation.OperationHelper.withConnection(OperationHelper.java:203)
	at com.mongodb.operation.CountOperation.execute(CountOperation.java:206)
	at com.mongodb.operation.CountOperation.execute(CountOperation.java:53)
	at com.mongodb.Mongo.execute(Mongo.java:773)
	at com.mongodb.Mongo$2.execute(Mongo.java:760)
	at com.mongodb.DBCollection.getCount(DBCollection.java:962)
Caused by: java.net.SocketException: Connection reset
	at java.net.SocketInputStream.read(SocketInputStream.java:210)
	at java.net.SocketInputStream.read(SocketInputStream.java:141)
	at com.mongodb.connection.SocketStream.read(SocketStream.java:85)
	at com.mongodb.connection.InternalStreamConnection.receiveResponseBuffers(InternalStreamConnection.java:491)
	at com.mongodb.connection.InternalStreamConnection.receiveMessage(InternalStreamConnection.java:221)
	at com.mongodb.connection.UsageTrackingInternalConnection.receiveMessage(UsageTrackingInternalConnection.java:102)
	at com.mongodb.connection.DefaultConnectionPool$PooledConnection.receiveMessage(DefaultConnectionPool.java:433)
	at com.mongodb.connection.CommandProtocol.execute(CommandProtocol.java:109)
	at com.mongodb.connection.DefaultServer$DefaultServerProtocolExecutor.execute(DefaultServer.java:159)
	at com.mongodb.connection.DefaultServerConnection.executeProtocol(DefaultServerConnection.java:286)
	at com.mongodb.connection.DefaultServerConnection.command(DefaultServerConnection.java:173)
	at com.mongodb.operation.CommandOperationHelper.executeWrappedCommandProtocol(CommandOperationHelper.java:215)
	at com.mongodb.operation.CommandOperationHelper.executeWrappedCommandProtocol(CommandOperationHelper.java:206)
	at com.mongodb.operation.CommandOperationHelper.executeWrappedCommandProtocol(CommandOperationHelper.java:112)
	at com.mongodb.operation.CountOperation$1.call(CountOperation.java:210)
	at com.mongodb.operation.CountOperation$1.call(CountOperation.java:206)
	at com.mongodb.operation.OperationHelper.withConnectionSource(OperationHelper.java:230)
	at com.mongodb.operation.OperationHelper.withConnection(OperationHelper.java:203)
	at com.mongodb.operation.CountOperation.execute(CountOperation.java:206)
	at com.mongodb.operation.CountOperation.execute(CountOperation.java:53)
2017-05-11 14:32:35,677 [application-akka.actor.default-dispatcher-3619] INFO  o.m.d.connection - Closed connection [connectionId{localValue:304}] to aws-eu-west-1-portal.0.dblayer.com:11301 because there was a socket exception raised on another connection from this pool.
2017-05-11 14:32:35,731 [application-akka.actor.default-dispatcher-3619] INFO  o.m.d.connection - Opened connection [connectionId{localValue:311}] to aws-eu-west-1-portal.0.dblayer.com:11301
2017-05-11 14:32:35,734 [cluster-ClusterId{value='5911d33fb03b837bba911406', description='null'}-aws-eu-west-1-portal.0.dblayer.com:11301] INFO  o.m.d.connection - Opened connection [connectionId{localValue:310}] to aws-eu-west-1-portal.0.dblayer.com:11301
2017-05-11 14:32:35,834 [application-akka.actor.default-dispatcher-3620] INFO  o.m.d.connection - Closed connection [connectionId{localValue:306}] to aws-eu-west-1-portal.0.dblayer.com:11301 because there was a socket exception raised on another connection from this pool.
2017-05-11 14:32:36,759 [deploy-akka.deploy-dispatcher-1471] DEBUG d.actors.TasksRunner - Sending completed message
2017-05-11 14:32:36,759 [deploy-akka.deploy-dispatcher-1471] DEBUG d.a.DeployGroupRunner - Deployment completed
```

--- a/magenta-lib/src/main/scala/magenta/graph/Graph.scala
+++ b/magenta-lib/src/main/scala/magenta/graph/Graph.scala
@@ -373,6 +373,20 @@ case class Graph[T](edges: Set[Edge[T]]) {
     }, identity)
   }
 
+  def allSuccesors(node: Node[T]): Set[Node[T]] = {
+    successors(node) match {
+      case s if s.isEmpty => Set.empty
+      case nodesSuccessors => nodesSuccessors ++ nodesSuccessors.flatMap(allSuccesors)
+    }
+  }
+
+  /** Returns a copy of the graph with all successor of the given node removed
+    */
+  def removeSuccessorValueNodes(node: ValueNode[T]): Graph[T] = {
+    val toBeRemoved = allSuccesors(node).filter(_ != EndNode)
+    Graph(edges.filter(edge => !(toBeRemoved.contains(edge.from) || toBeRemoved.contains(edge.to))) + Edge(node, EndNode))
+  }
+
   lazy val toList: List[T] = nodeList.flatMap(_.maybeValue)
 
   override def toString: String = {

--- a/magenta-lib/src/test/scala/magenta/graph/GraphTest.scala
+++ b/magenta-lib/src/test/scala/magenta/graph/GraphTest.scala
@@ -259,4 +259,9 @@ class GraphTest extends FlatSpec with ShouldMatchers {
       one ~> three, one ~2~> four, two ~> three, two ~2~> four
     )
   }
+
+  "removeSuccesors" should "provide a copy of the graph with successors to a given node removed" in {
+    val graph: Graph[Int] = Graph.from(Seq(1, 2, 3)).joinParallel(Graph.from(Seq(4, 5)))
+    graph.removeSuccessorValueNodes(ValueNode(1)) shouldBe Graph.from(Seq(1)).joinParallel(Graph.from(Seq(4, 5)))
+  }
 }

--- a/riff-raff/app/deployment/actors/DeployGroupRunner.scala
+++ b/riff-raff/app/deployment/actors/DeployGroupRunner.scala
@@ -47,15 +47,17 @@ class DeployGroupRunner(
   var completed: Set[ValueNode[DeploymentTasks]] = Set.empty
   var failed: Set[ValueNode[DeploymentTasks]] = Set.empty
 
-  def deploymentGraph: Graph[DeploymentTasks] = failed.foldLeft(deployContext.map(_.tasks).getOrElse(Graph.empty[DeploymentTasks])) {
+  def deploymentGraph: Graph[DeploymentTasks] = deployContext.map(_.tasks).getOrElse(Graph.empty[DeploymentTasks])
+
+  def reachableDeploymentGraph: Graph[DeploymentTasks] = failed.foldLeft(deploymentGraph) {
     (graph, failedNode) => graph.removeSuccessorValueNodes(failedNode)
   }
-  def allDeployments = deploymentGraph.nodes.filterValueNodes
+  def reachableDeployments = reachableDeploymentGraph.nodes.filterValueNodes
 
-  def isFinished: Boolean = allDeployments == completed ++ failed
+  def isFinished: Boolean = reachableDeployments == completed ++ failed
   def isExecuting: Boolean = executing.nonEmpty
 
-  def first: List[ValueNode[DeploymentTasks]] = deploymentGraph.orderedSuccessors(StartNode).filterValueNodes
+  def first: List[ValueNode[DeploymentTasks]] = reachableDeploymentGraph.orderedSuccessors(StartNode).filterValueNodes
   /* these two functions can return a number of things
       - Deployments: list of deployments
       - FinishPath: indicator there are no more tasks on this path
@@ -67,9 +69,9 @@ class DeployGroupRunner(
     // otherwise let's see what children are valid to return
     else {
       // candidates are all successors not already executing or completing
-      val nextDeploymentCandidates = deploymentGraph.orderedSuccessors(deploymentGraph.get(deployment)).filterValueNodes
+      val nextDeploymentCandidates = reachableDeploymentGraph.orderedSuccessors(reachableDeploymentGraph.get(deployment)).filterValueNodes
       // now filter for only tasks whose predecessors are all completed
-      val nextDeployments = nextDeploymentCandidates.filter { deployment => (deploymentGraph.predecessors(deployment) -- completed).isEmpty }
+      val nextDeployments = nextDeploymentCandidates.filter { deployment => (reachableDeploymentGraph.predecessors(deployment) -- completed).isEmpty }
       if (nextDeployments.nonEmpty) {
         NextTasks(nextDeployments)
       } else {
@@ -115,7 +117,7 @@ class DeployGroupRunner(
   override def toString: String = {
     s"""
        |UUID: $id
-       |#Tasks: ${allDeployments.size}
+       |#Tasks: ${reachableDeployments.size}
        |#Executing: ${executing.mkString("; ")}
        |#Completed: ${completed.size} Failed: ${failed.size}
        |#Done: ${completed.size+failed.size}

--- a/riff-raff/app/deployment/actors/DeployGroupRunner.scala
+++ b/riff-raff/app/deployment/actors/DeployGroupRunner.scala
@@ -47,7 +47,9 @@ class DeployGroupRunner(
   var completed: Set[ValueNode[DeploymentTasks]] = Set.empty
   var failed: Set[ValueNode[DeploymentTasks]] = Set.empty
 
-  def deploymentGraph: Graph[DeploymentTasks] = deployContext.map(_.tasks).getOrElse(Graph.empty[DeploymentTasks])
+  def deploymentGraph: Graph[DeploymentTasks] = failed.foldLeft(deployContext.map(_.tasks).getOrElse(Graph.empty[DeploymentTasks])) {
+    (graph, failedNode) => graph.removeSuccessorValueNodes(failedNode)
+  }
   def allDeployments = deploymentGraph.nodes.filterValueNodes
 
   def isFinished: Boolean = allDeployments == completed ++ failed

--- a/riff-raff/test/deployment/Fixtures.scala
+++ b/riff-raff/test/deployment/Fixtures.scala
@@ -27,6 +27,10 @@ object Fixtures extends MockitoSugar {
     DeploymentGraph(twoTasks, "branch one") joinParallel DeploymentGraph(twoTasks, "branch two")
   }
 
+  val dependentGraph: Graph[DeploymentTasks] = {
+    (DeploymentGraph(twoTasks, "one") joinSeries DeploymentGraph(twoTasks, "two")) joinParallel DeploymentGraph(twoTasks, "branch two")
+  }
+
   def createRecord(
     projectName: String = "test",
     stage: String = "TEST",

--- a/riff-raff/test/deployment/actors/DeployGroupRunnerTest.scala
+++ b/riff-raff/test/deployment/actors/DeployGroupRunnerTest.scala
@@ -18,7 +18,7 @@ class DeployGroupRunnerTest extends TestKit(ActorSystem("DeployGroupRunnerTest")
   "DeployGroupRunnerTest" should "initalise the state from a set of tasks" in {
     val dr = createDeployRunnerWithUnderlying()
     prepare(dr, threeSimpleTasks)
-    dr.ul.allDeployments.size should be(1)
+    dr.ul.reachableDeployments.size should be(1)
     dr.ul.isExecuting should be(false)
   }
 

--- a/riff-raff/test/deployment/actors/DeployGroupRunnerTest.scala
+++ b/riff-raff/test/deployment/actors/DeployGroupRunnerTest.scala
@@ -94,6 +94,23 @@ class DeployGroupRunnerTest extends TestKit(ActorSystem("DeployGroupRunnerTest")
     dr.deployCoordinatorProbe.expectMsgClass(classOf[DeployCoordinator.CleanupDeploy])
   }
 
+  it should "be finished if all tasks that could potentially be run (given failures) have completed" in {
+    val dr = createDeployRunnerWithUnderlying()
+    prepare(dr, dependentGraph)
+    dr.ref ! DeployGroupRunner.StartDeployment
+    val runDeployment = dr.deploymentRunnerProbe.expectMsgClass(classOf[TasksRunner.RunDeployment])
+    val secondTaskSet = dr.deploymentRunnerProbe.expectMsgClass(classOf[TasksRunner.RunDeployment])
+    dr.deploymentRunnerProbe.reply(DeployGroupRunner.DeploymentFailed(runDeployment.deployment, new RuntimeException("test exception")))
+    dr.ul.isExecuting should be(true)
+    dr.ul.failed should be(Set(ValueNode(runDeployment.deployment)))
+    dr.ul.isFinished should be(false)
+    dr.deploymentRunnerProbe.reply(DeployGroupRunner.DeploymentCompleted(secondTaskSet.deployment))
+    dr.ul.isExecuting should be (false)
+    dr.ul.isFinished should be(true)
+    dr.deploymentRunnerProbe.expectNoMsg()
+    dr.deployCoordinatorProbe.expectMsgClass(classOf[DeployCoordinator.CleanupDeploy])
+  }
+
   val deploymentTypes = Nil
 
   trait DR {

